drivers/timer/cortex_m_systick.c

```diff
@@ -88,13 +88,18 @@ void z_clock_isr(void *arg)
 	ARG_UNUSED(arg);
 	u32_t dticks;
 
+	elapsed();
+
+#if !IS_ENABLED(CONFIG_QEMU_TICKLESS_WORKAROUND)
+	cycle_count += overflow_cyc;
+#else
+	/* QEMU will send IRQ, but not sure about setting SysTick->CTRL */
 	cycle_count += last_load;
+#endif
+	overflow_cyc = 0;
```

Comment:
This isn't a new problem, but I'm concerned about a race condition here: this mutates `overflow_cyc` without holding the spinlock.  If a higher priority interrupt invokes `elapsed()` through any other function (like `z_clock_elapsed()` or `z_timer_cycle_get_32()` or `z_clock_set_timeout()`) information may be lost.

Reply:
@pabigot you are right, as all other uses of `elapsed()` are guarded by spinlocks, i've added an irq lock so no other ISR would also call `elapsed()`. 
I was debugging #18282 and what @andyross suggested (blocking multiple irq interrupts by using a busy wait). Looks like guarding the variable changes in `elapsed()` makes the behavior coherent.